bundles/io/org.eclipse.smarthome.io.transport.serial/src/main/java/org/eclipse/smarthome/io/transport/serial/SerialPort.java

```diff
@@ -0,0 +1,159 @@
+     * If the port is unidirectional and doesn't support receiving data, then getInputStream returns null.
+     *
+     * <p>
+     * The read behaviour of the input stream returned by getInputStream depends on combination of the threshold and
+     * timeout values. The possible behaviours are described in the table below:
+     *
+     * | Threshold .........| Timeout ........ | Read Buffer Size | Read Behaviour |
+     * | State ...| Value ..| State ...| Value |
+     * | disabled | - ......| disabled | - ....| n bytes .........| block until any data is available
+     * | enabled .| m bytes | disabled | - ....| n bytes .........| block until min(m,n) bytes are available
+     * | disabled | - ......| enabled .| x ms .| n bytes .........| block for x ms or until any data is available
+     * | enabled .| m bytes | enabled .| x ms .| n bytes .........| block for x ms or until min(m,n) bytes are available
+     *
+     * <p>
+     * Note, however, that framing errors may cause the Timeout and Threshold values to complete prematurely without
+     * raising an exception.
+     *
+     * <p>
+     * Enabling the Timeout OR Threshold with a value a zero is a special case. This causes the underlying driver to
+     * poll for incoming data instead being event driven. Otherwise, the behaviour is identical to having both the
+     * Timeout and Threshold disabled.
+     * *
+     *
+     * @return the input stream or null
+     * @throws IOException on I/O error
+     */
+    @Nullable
+    InputStream getInputStream() throws IOException;
+
+    /**
+     * Returns an output stream.
+     *
+     * <p>
+     * This is the only way to send data to the communications port.
+     * If the port is unidirectional and doesn't support sending data, then getOutputStream returns null.
+     *
+     * @return the output stream or null
+     * @throws IOException on I/O error
+     */
+    @Nullable
+    OutputStream getOutputStream() throws IOException;
+
+    /**
+     * Registers a {@link SerialPortEventListener} object to listen for {@link SerialEvents}.
+     *
+     * <p>
+     * Interest in specific events may be expressed using the notifyOnXXX calls.
+     * The serialEvent method of SerialPortEventListener will be called with a SerialEvent object describing the event.
+     *
+     * Only one listener per SerialPort is allowed.
+     * Once a listener is registered, subsequent call attempts to addEventListener will throw a
+     * TooManyListenersException without effecting the listener already registered.
+     *
+     * <p>
+     * All the events received by this listener are generated by one dedicated thread that belongs to the SerialPort
+     * object.
+     * After the port is closed, no more event will be generated.
+     *
+     * @param listener the listener
+     * @throws TooManyListenersException if too many listeners has been added
+     */
+    void addEventListener(SerialPortEventListener listener) throws TooManyListenersException;
```

Comment:
annotation on listener denoting it cannot be null

Reply:
Have you realized that the whole class is annotated by `NonNullByDefault`? No need to add the `NonNull` annotation to any members, only `Nullable` would be needed.

Reply:
Didn't notice that at all. Sorry for the noise. 👍